File: sdk/typescript/tools.ts
/**
 * LLM SDK - Tool Use / Function Calling
 * =====================================
 * Enable LLM to call business functions
 *
 * Usage:
 *   import { client } from './index'
 *   import { llmTool, ChatWithToolsMixin } from './tools'
 *
 *   const getWeather = llmTool({
 *     name: 'get_weather',
 *     description: 'Get weather for a city',
 *     parameters: {
 *       type: 'object',
 *       properties: {
 *         city: { type: 'string', description: 'City name' }
 *       },
 *       required: ['city']
 *     }
 *   })(async (city: string) => {
 *     return await weatherApi.fetch(city)
 *   })
 *
 *   // Direct call
 *   const reply = await client.chatWithTools('Beijing weather?', [getWeather])
 *
 *   // Or extend client
 *   class MyClient extends ChatWithToolsMixin(LLMClient) {}
 */

import { z } from 'zod'
import { zodToJsonSchema } from 'zod-to-json-schema'

// Re-export zod types for convenience
export { z }

// Tool definition interface
export interface ToolDefinition<T extends (...args: any[]) => any = any> {
  name: string
  description: string
  parameters: Record<string, any>
  execute: T
}

// Create a tool from a function with Zod schema
export function llmTool(
  options: {
    name?: string
    description?: string
    parameters?: Record<string, any>
    schema?: z.ZodType
  },
  execute?: any
): (fn: any) => any {
  return (fn: any): any => {
    const name = options.name || fn.name
    let description = options.description || ''
    let parameters = options.parameters || { type: 'object', properties: {} }

    // If Zod schema provided, convert to JSON Schema
    if (options.schema) {
      parameters = (zodToJsonSchema as any)(options.schema, { target: 'openApi3' })
    }

    // Try to get description from function
    if (!description && (fn as any).description) {
      description = (fn as any).description
    }

    const propertyNames = Object.keys((parameters as any)?.properties ?? {})
    const executor = execute || ((args: Record<string, any>) => {
      if (propertyNames.length === 0) {
        return fn(args)
      }
      if (propertyNames.length === 1 && fn.length <= 1) {
        return fn(args[propertyNames[0]])
      }
      return fn(...propertyNames.map(key => args[key]))
    })

    return {
      name,
      description,
      parameters,
      execute: executor,
    }
  }
}

// Decorator style (for class methods)
export function tool(
  name?: string,
  description?: string,
  parameters?: Record<string, any>,
) {
  return function (target: any, propertyKey: string, descriptor: PropertyDescriptor) {
    const originalMethod = descriptor.value
    descriptor.value = function (...args: any[]) {
      return originalMethod.apply(this, args)
    }
    // Store tool metadata
    ;(descriptor.value as any).toolMeta = {
      name: name || propertyKey,
      description: description || '',
      parameters: parameters || { type: 'object', properties: {} },
    }
    return descriptor
  }
}

// Tool call result
export class ToolCall {
  constructor(
    public toolCallId: string,
    public name: string,
    public args: Record<string, any>,
  ) {}

  async execute(tools: Map<string, ToolDefinition>): Promise<string> {
    const tool = tools.get(this.name)
    if (!tool) {
      return `Error: Unknown tool '${this.name}'`
    }

    try {
      const result = await tool.execute(this.args)
      if (typeof result === 'object' && result !== null) {
        return JSON.stringify(result)
      }
      return String(result)
    } catch (e) {
      return `Error: ${e instanceof Error ? e.message : String(e)}`
    }
  }

  toMessage() {
    return {
      id: this.toolCallId,
      type: 'function',
      function: {
        name: this.name,
        arguments: JSON.stringify(this.args),
      },
    }
  }
}

function extractToolCalls(content: string, rawToolCalls?: any[]): ToolCall[] {
  if (rawToolCalls?.length) {
    return rawToolCalls.map(
      (toolCall: any) =>
        new ToolCall(
          toolCall.id,
          toolCall.function.name,
          JSON.parse(toolCall.function.arguments || '{}'),
        ),
    )
  }

  try {
    const match = content.match(/\[[\s\S]*\]/)
    if (!match) {
      return []
    }
    const parsed = JSON.parse(match[0])
    if (!Array.isArray(parsed)) {
      return []
    }
    return parsed
      .filter((c: any) => c && c.name)
      .map(
        (c: any, i: number) =>
          new ToolCall(
            String(i),
            c.name,
            c.arguments || {},
          ),
      )
  } catch {
    return []
  }
}

export async function chatWithTools(
  client: any,
  prompt: string,
  tools: ToolDefinition[],
  opts: {
    system?: string
    history?: any[]
    capability?: string
    model?: string
    maxToolCalls?: number
    imageUrl?: string
    imageBase64?: { data: string; mime?: string }
    maxTokens?: number
    temperature?: number
    [key: string]: any
  } = {},
): Promise<string> {
  const maxToolCalls = opts.maxToolCalls || 10
  const toolMap = new Map<string, ToolDefinition>()
  for (const tool of tools) {
    toolMap.set(tool.name, tool)
  }

  const { maxToolCalls: _maxToolCalls, ...requestOpts } = opts
  const hasImage = !!(requestOpts.imageUrl || requestOpts.imageBase64)
  const model = client.resolveModel('chat', hasImage, requestOpts)
  const messages = client.buildMessages(prompt, requestOpts)

  for (let i = 0; i < maxToolCalls; i++) {
    const response = await client.openai.chat.completions.create({
      model,
      messages,
      tools: tools.map(t => toolToOpenAISchema(t)),
      tool_choice: 'auto',
      ...(requestOpts.maxTokens ? { max_tokens: requestOpts.maxTokens } : {}),
      ...(requestOpts.temperature !== undefined ? { temperature: requestOpts.temperature } : {}),
    })

    const message = response.choices[0]?.message
    const content = message?.content || ''
    const toolCalls = extractToolCalls(content, message?.tool_calls as any[])
    const assistantMessage: any = { role: 'assistant', content }
    if (toolCalls.length) {
      assistantMessage.tool_calls = toolCalls.map(tc => tc.toMessage())
    }
    messages.push(assistantMessage)

    if (!toolCalls.length) {
      return content
    }

    for (const tc of toolCalls) {
      const result = await tc.execute(toolMap)
      messages.push({
        role: 'tool',
        tool_call_id: tc.toolCallId,
        content: result,
      })
    }
  }

  return 'Tool call limit reached'
}

export function ChatWithToolsMixin<T extends new (...args: any[]) => any>(Client: T) {
  return class extends Client {
    async chatWithTools(
      prompt: string,
      tools: ToolDefinition[],
      opts: {
        system?: string
        history?: any[]
        capability?: string
        model?: string
        maxToolCalls?: number
        [key: string]: any
      } = {},
    ): Promise<string> {
      return chatWithTools(this, prompt, tools, opts)
    }
  } as any
}

// Helper to convert to OpenAI schema format
export function toolToOpenAISchema(tool: ToolDefinition): any {
  return {
    type: 'function',
    function: {
      name: tool.name,
      description: tool.description,
      parameters: tool.parameters,
    },
  }
}

File: sdk/go/tools.go
// Package llmclient provides Tool Use / Function Calling support.
//
// Usage:
//
//	type WeatherParams struct {
//		City string `json:"city" description:"城市名称"`
//	}
//
//	type WeatherTool struct{}
//
//	func (w WeatherTool) Name() string { return "get_weather" }
//	func (w WeatherTool) Description() string { return "获取城市天气" }
//	func (w WeatherTool) Parameters() interface{} { return WeatherParams{} }
//	func (w WeatherTool) Execute(ctx context.Context, params interface{}) (string, error) {
//		p := params.(WeatherParams)
//		return weatherAPI.Fetch(p.City), nil
//	}
//
//	// Use with client
//	err := c.ChatWithTools(ctx, "北京天气如何?", &WeatherTool{}, nil)
package llmclient

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"reflect"
)

// Tool is the interface that all tools must implement.
type Tool interface {
	Name() string
	Description() string
	Parameters() interface{}
	Execute(ctx context.Context, params interface{}) (string, error)
}

// toolCall represents a single tool call from the LLM.
type toolCall struct {
	Index    int
	ID      string
	Name    string
	Args    map[string]interface{}
}

// ToolDefinition converts a Tool to OpenAI function calling format.
func ToolDefinition(t Tool) map[string]interface{} {
	return map[string]interface{}{
		"type": "function",
		"function": map[string]interface{}{
			"name":        t.Name(),
			"description": t.Description(),
			"parameters":  jsonSchema(t.Parameters()),
		},
	}
}

// jsonSchema converts a Go struct to JSON Schema.
func jsonSchema(v interface{}) map[string]interface{} {
	if v == nil {
		return map[string]interface{}{"type": "object"}
	}

	t := reflect.TypeOf(v)
	if t.Kind() == reflect.Ptr {
		t = t.Elem()
	}
	if t.Kind() != reflect.Struct {
		return map[string]interface{}{"type": "object"}
	}

	properties := make(map[string]interface{})
	required := []string{}

	for i := 0; i < t.NumField(); i++ {
		field := t.Field(i)
		jsonTag := field.Tag.Get("json")
		descTag := field.Tag.Get("description")

		// Parse json tag (e.g., "city,omitempty")
		name := field.Name
		if jsonTag != "" {
			if n, _, _ := parseTag(jsonTag); n != "" {
				name = n
			}
		}

		// Field type to JSON Schema type
		fieldType := goTypeToJSON(field.Type)

		prop := map[string]interface{}{
			"type": fieldType,
		}
		if descTag != "" {
			prop["description"] = descTag
		}

		properties[name] = prop

		// Check if required (no omitempty)
		if jsonTag == "" || !containsOmitEmpty(jsonTag) {
			required = append(required, name)
		}
	}

	result := map[string]interface{}{
		"type":       "object",
		"properties": properties,
	}
	if len(required) > 0 {
		result["required"] = required
	}

	return result
}

func parseTag(tag string) (name string, omitEmpty bool, anyOther bool) {
	parts := splitTag(tag)
	if len(parts) == 0 {
		return "", false, false
	}
	name = parts[0]
	for _, p := range parts[1:] {
		switch p {
		case "omitempty":
			omitEmpty = true
		default:
			anyOther = true
		}
	}
	return name, omitEmpty, anyOther
}

func containsOmitEmpty(tag string) bool {
	return len(tag) > 0 && (tag == "omitempty" || len(tag) > 10 && tag[:10] == "omitempty")
}

func splitTag(s string) []string {
	var result []string
	var current []rune
	for _, r := range s {
		if r == ',' {
			result = append(result, string(current))
			current = nil
		} else {
			current = append(current, r)
		}
	}
	if len(current) > 0 {
		result = append(result, string(current))
	}
	return result
}

func goTypeToJSON(t reflect.Type) string {
	if t == nil {
		return "string"
	}

	// Handle pointers
	if t.Kind() == reflect.Ptr {
		return goTypeToJSON(t.Elem())
	}

	switch t.Kind() {
	case reflect.Bool:
		return "boolean"
	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
		return "integer"
	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
		return "integer"
	case reflect.Float32, reflect.Float64:
		return "number"
	case reflect.String:
		return "string"
	case reflect.Slice, reflect.Array:
		return "array"
	case reflect.Map:
		return "object"
	case reflect.Struct:
		return "object"
	default:
		return "string"
	}
}

// ChatWithTools executes a chat with tool calling capability.
// It automatically handles the tool call loop: LLM requests tool -> execute -> return result -> continue.
func (c *Client) ChatWithTools(ctx context.Context, prompt string, tool Tool, opts *ChatOpts) error {
	if opts == nil {
		opts = &ChatOpts{}
	}
	toolDef := ToolDefinition(tool)
	model, err := c.resolveOpts(opts, opts.ImageURL != "")
	if err != nil {
		return err
	}
	baseMessages := c.buildMessages(prompt, opts)
	messages := make([]map[string]interface{}, 0, len(baseMessages))
	for _, msg := range baseMessages {
		messages = append(messages, map[string]interface{}{
			"role":    msg.Role,
			"content": msg.Content,
		})
	}

	maxCalls := 10
	for i := 0; i < maxCalls; i++ {
		body := map[string]interface{}{
			"model":    model,
			"messages": messages,
			"tools":    []map[string]interface{}{toolDef},
			"tool_choice": "auto",
		}

		// Add other options
		if opts.Temperature != nil {
			body["temperature"] = *opts.Temperature
		}
		if opts.MaxTokens > 0 {
			body["max_tokens"] = opts.MaxTokens
		}

		var respData struct {
			Choices []struct {
				Message struct {
					Content     string `json:"content"`
					ToolCalls   []struct {
						ID       string          `json:"id"`
						Type     string          `json:"type"`
						Function struct {
							Name      string          `json:"name"`
							Arguments json.RawMessage `json:"arguments"`
						} `json:"function"`
					} `json:"tool_calls"`
				} `json:"message"`
			} `json:"choices"`
		}

		resp, err := c.postJSON(ctx, "/v1/chat/completions", body)
		if err != nil {
			return err
		}
		raw, _ := io.ReadAll(resp.Body)
		resp.Body.Close()
		if resp.StatusCode >= 400 {
			return classifyError(resp.StatusCode, string(raw))
		}
		if err := json.Unmarshal(raw, &respData); err != nil {
			return err
		}

		if len(respData.Choices) == 0 {
			return fmt.Errorf("no response from LLM")
		}

		msg := respData.Choices[0].Message
		assistantMessage := map[string]interface{}{
			"role":    "assistant",
			"content": msg.Content,
		}
		if len(msg.ToolCalls) > 0 {
			toolCalls := make([]map[string]interface{}, 0, len(msg.ToolCalls))
			for _, tc := range msg.ToolCalls {
				arguments := string(tc.Function.Arguments)
				var encoded string
				if err := json.Unmarshal(tc.Function.Arguments, &encoded); err == nil {
					arguments = encoded
				}
				toolCalls = append(toolCalls, map[string]interface{}{
					"id":   tc.ID,
					"type": tc.Type,
					"function": map[string]interface{}{
						"name":      tc.Function.Name,
						"arguments": arguments,
					},
				})
			}
			assistantMessage["tool_calls"] = toolCalls
		}
		messages = append(messages, assistantMessage)

		if len(msg.ToolCalls) == 0 {
			return nil
		}

		for _, tc := range msg.ToolCalls {
			if tc.Type != "function" {
				continue
			}

			var args map[string]interface{}
			if err := json.Unmarshal(tc.Function.Arguments, &args); err != nil {
				var encoded string
				if err := json.Unmarshal(tc.Function.Arguments, &encoded); err == nil {
					_ = json.Unmarshal([]byte(encoded), &args)
				}
				if args == nil {
					args = make(map[string]interface{})
				}
			}

			params, err := decodeToolArguments(tool.Parameters(), args)
			if err != nil {
				params = args
			}
			result, err := tool.Execute(ctx, params)
			if err != nil {
				result = fmt.Sprintf("Error: %v", err)
			}

			messages = append(messages, map[string]interface{}{
				"role":         "tool",
				"tool_call_id": tc.ID,
				"content":      result,
			})
		}
	}

	return fmt.Errorf("tool call limit reached")
}

func decodeToolArguments(template interface{}, args map[string]interface{}) (interface{}, error) {
	if template == nil {
		return args, nil
	}
	raw, err := json.Marshal(args)
	if err != nil {
		return nil, err
	}
	targetType := reflect.TypeOf(template)
	if targetType.Kind() == reflect.Ptr {
		value := reflect.New(targetType.Elem())
		if err := json.Unmarshal(raw, value.Interface()); err != nil {
			return nil, err
		}
		return value.Interface(), nil
	}
	value := reflect.New(targetType)
	if err := json.Unmarshal(raw, value.Interface()); err != nil {
		return nil, err
	}
	return value.Elem().Interface(), nil
}
